File: ios/CarbBookCore/Sources/CarbBookCore/Carbs.swift
import Foundation

/// Mirrors packages/core/src/carbs.ts.
public protocol Catalog: Sendable {
    func food(_ id: Id) -> FoodData?
    func portions(_ foodId: Id) -> [PortionData]
    func meal(_ id: Id) -> MealData?
    func mealItems(_ mealId: Id) -> [MealItemData]
}

public struct CarbResult: Equatable, Sendable {
    public var carbsG: Double
    public var complete: Bool

    public init(carbsG: Double, complete: Bool) {
        self.carbsG = carbsG
        self.complete = complete
    }

    public static let incomplete = CarbResult(carbsG: 0, complete: false)
}

public struct InMemoryCatalog: Catalog {
    private let foods: [Id: FoodData]
    private let portionsByFood: [Id: [PortionData]]
    private let meals: [Id: MealData]
    private let itemsByMeal: [Id: [MealItemData]]

    /// Rows with `deleted == 1` are skipped. Meal items are ordered by `position`, keeping input
    /// order for equal positions (JS `Array.prototype.sort` is stable).
    public init(foods: [FoodData] = [], portions: [PortionData] = [], meals: [MealData] = [], mealItems: [MealItemData] = []) {
        var foodMap: [Id: FoodData] = [:]
        for f in foods where f.deleted != 1 { foodMap[f.id] = f }
        var mealMap: [Id: MealData] = [:]
        for m in meals where m.deleted != 1 { mealMap[m.id] = m }
        var portionMap: [Id: [PortionData]] = [:]
        for p in portions where p.deleted != 1 { portionMap[p.foodId, default: []].append(p) }
        var itemMap: [Id: [(Int, MealItemData)]] = [:]
        for (index, item) in mealItems.enumerated() where item.deleted != 1 {
            itemMap[item.mealId, default: []].append((index, item))
        }
        self.foods = foodMap
        self.meals = mealMap
        self.portionsByFood = portionMap
        self.itemsByMeal = itemMap.mapValues { list in
            list.sorted { $0.1.position != $1.1.position ? $0.1.position < $1.1.position : $0.0 < $1.0 }.map(\.1)
        }
    }

    public func food(_ id: Id) -> FoodData? { foods[id] }
    public func portions(_ foodId: Id) -> [PortionData] { portionsByFood[foodId] ?? [] }
    public func meal(_ id: Id) -> MealData? { meals[id] }
    public func mealItems(_ mealId: Id) -> [MealItemData] { itemsByMeal[mealId] ?? [] }
}

/// Mass path: carbs_per_100g if valid. Only a nil carbs_per_100g falls back to carbs_per_100ml +
/// density; a present-but-invalid value fails closed (incomplete).
private func carbsForGrams(_ food: FoodData, _ portions: [PortionData], _ grams: Double) -> CarbResult {
    if food.carbsPer100g != nil {
        return isValidCarbsPer100g(food.carbsPer100g)
            ? CarbResult(carbsG: grams * food.carbsPer100g! / 100, complete: true)
            : .incomplete
    }
    if isValidCarbsPer100ml(food.carbsPer100ml), let density = densityOf(food, portions) {
        return CarbResult(carbsG: (grams / density) * food.carbsPer100ml! / 100, complete: true)
    }
    return .incomplete
}

private func foodItemCarbs(_ catalog: Catalog, _ foodId: Id, _ amount: Double, _ unit: String) -> CarbResult {
    guard let food = catalog.food(foodId), isValidAmount(amount) else { return .incomplete }
    let portions = catalog.portions(foodId)
    if isVolumeUnit(unit) {
        if food.carbsPer100ml != nil {
            // Present per-100 ml basis: use it or fail closed; never fall back to density.
            guard isValidCarbsPer100ml(food.carbsPer100ml) else { return .incomplete }
            return CarbResult(carbsG: amount * Units.volumeMl[unit]! * food.carbsPer100ml! / 100, complete: true)
        }
        // Nil per-100 ml: per-100 g + density only.
        guard let grams = foodAmountToGrams(amount, unit, food, portions),
              isValidCarbsPer100g(food.carbsPer100g) else { return .incomplete }
        return CarbResult(carbsG: grams * food.carbsPer100g! / 100, complete: true)
    }
    if unit.hasPrefix(Units.portionPrefix) {
        let portionId = String(unit.dropFirst(Units.portionPrefix.count))
        guard let portion = portions.first(where: { $0.id == portionId }) else { return .incomplete }
        // carbs_g on volume portions is ignored (grams path only).
        if portion.kind != "volume", portion.carbsG != nil {
            // Present portion carbs: use them or fail closed; never fall back to grams.
            guard isValidPortionCarbs(portion.carbsG), portion.quantity.isFinite, portion.quantity > 0 else {
                return .incomplete
            }
            return CarbResult(carbsG: (amount / portion.quantity) * portion.carbsG!, complete: true)
        }
    }
    // Mass units, and portions without carbs (or volume portions), go through a known weight.
    guard let grams = foodAmountToGrams(amount, unit, food, portions) else { return .incomplete }
    return carbsForGrams(food, portions, grams)
}

private func mealTotalCarbs(_ catalog: Catalog, _ mealId: Id, _ visiting: inout Set<Id>) -> CarbResult {
    if visiting.contains(mealId) || catalog.meal(mealId) == nil { return .incomplete }
    visiting.insert(mealId)
    let items = catalog.mealItems(mealId)
    var results: [CarbResult] = []
    for item in items {
        results.append(resolveItem(catalog, item.refType, item.refId, item.amount, item.unit, &visiting))
    }
    let total = items.isEmpty ? .incomplete : sumCarbs(results)
    visiting.remove(mealId)
    return total
}

private func mealItemCarbs(_ catalog: Catalog, _ mealId: Id, _ amount: Double, _ unit: String, _ visiting: inout Set<Id>) -> CarbResult {
    guard let meal = catalog.meal(mealId), amount.isFinite, amount >= 0 else { return .incomplete }
    var factor: Double?
    if unit == Units.serving && meal.yieldServings.isFinite && meal.yieldServings > 0 {
        factor = amount / meal.yieldServings
    } else if let grams = Units.massGrams[unit], let weight = meal.totalWeightG, weight.isFinite, weight > 0 {
        factor = amount * grams / weight
    }
    guard let factor else { return .incomplete }
    let total = mealTotalCarbs(catalog, mealId, &visiting)
    return CarbResult(carbsG: total.carbsG * factor, complete: total.complete)
}

private func resolveItem(_ catalog: Catalog, _ refType: RefType, _ refId: Id, _ amount: Double, _ unit: String,
                         _ visiting: inout Set<Id>) -> CarbResult {
    switch refType {
    case .food: foodItemCarbs(catalog, refId, amount, unit)
    case .meal: mealItemCarbs(catalog, refId, amount, unit, &visiting)
    }
}

/// Carbs for one line item (a food or a meal) at the given amount and unit.
public func itemCarbs(_ catalog: Catalog, _ refType: RefType, _ refId: Id, _ amount: Double, _ unit: String) -> CarbResult {
    var visiting = Set<Id>()
    return resolveItem(catalog, refType, refId, amount, unit, &visiting)
}

public func sumCarbs(_ results: [CarbResult]) -> CarbResult {
    results.reduce(CarbResult(carbsG: 0, complete: true)) {
        CarbResult(carbsG: $0.carbsG + $1.carbsG, complete: $0.complete && $1.complete)
    }
}

/// True if adding `candidateMealId` as a component of `mealId` would make a meal contain itself.
public func wouldCreateCycle(_ catalog: Catalog, _ mealId: Id, _ candidateMealId: Id) -> Bool {
    if candidateMealId == mealId { return true }
    var stack = [candidateMealId]
    var seen = Set<Id>()
    while let current = stack.popLast() {
        if seen.contains(current) { continue }
        seen.insert(current)
        for item in catalog.mealItems(current) where item.refType == .meal {
            if item.refId == mealId { return true }
            stack.append(item.refId)
        }
    }
    return false
}

File: packages/core/src/carbs.ts
import type { FoodData, Id, MealData, MealItemData, PortionData, RefType } from './types';
import {
  MASS_UNITS,
  PORTION_PREFIX,
  SERVING_UNIT,
  VOLUME_UNITS,
  densityOf,
  foodAmountToGrams,
  isMassUnit,
  isValidAmount,
  isValidCarbsPer100g,
  isValidCarbsPer100ml,
  isValidPortionCarbs,
  isVolumeUnit,
} from './units';

export interface Catalog {
  food(id: Id): FoodData | undefined;
  portions(foodId: Id): PortionData[];
  meal(id: Id): MealData | undefined;
  mealItems(mealId: Id): MealItemData[];
}

export interface CarbResult {
  carbs_g: number;
  complete: boolean;
}

const INCOMPLETE: CarbResult = { carbs_g: 0, complete: false };

function groupBy<T>(rows: T[], key: (row: T) => Id): Map<Id, T[]> {
  const map = new Map<Id, T[]>();
  for (const row of rows) {
    const k = key(row);
    const list = map.get(k);
    if (list) list.push(row);
    else map.set(k, [row]);
  }
  return map;
}

type Deletable<T> = T & { deleted?: 0 | 1 };

function notDeleted<T extends { deleted?: 0 | 1 }>(rows: T[]): T[] {
  return rows.filter((row) => row.deleted !== 1);
}

export function createCatalog(data: {
  foods?: Deletable<FoodData>[];
  portions?: Deletable<PortionData>[];
  meals?: Deletable<MealData>[];
  meal_items?: Deletable<MealItemData>[];
}): Catalog {
  const foods = new Map(notDeleted(data.foods ?? []).map((f) => [f.id, f]));
  const meals = new Map(notDeleted(data.meals ?? []).map((m) => [m.id, m]));
  const portions = groupBy(notDeleted(data.portions ?? []), (p) => p.food_id);
  const items = groupBy(notDeleted(data.meal_items ?? []), (i) => i.meal_id);
  for (const list of items.values()) list.sort((a, b) => a.position - b.position);
  return {
    food: (id) => foods.get(id),
    portions: (foodId) => portions.get(foodId) ?? [],
    meal: (id) => meals.get(id),
    mealItems: (mealId) => items.get(mealId) ?? [],
  };
}

/**
 * Fail closed (any-unit foods addendum): only an absent (null/undefined) basis may fall back to
 * another path. A present-but-invalid basis (out of range, negative, NaN) makes that unit family
 * incomplete. Keep foodUnits in units.ts in step with this.
 */

/** Mass path: carbs_per_100g if present; else carbs_per_100ml + density if present; else incomplete. */
function carbsForGrams(food: FoodData, portions: PortionData[], grams: number): CarbResult {
  if (food.carbs_per_100g != null) {
    return isValidCarbsPer100g(food.carbs_per_100g)
      ? { carbs_g: (grams * food.carbs_per_100g) / 100, complete: true }
      : INCOMPLETE;
  }
  if (food.carbs_per_100ml != null) {
    const density = densityOf(food, portions);
    if (!isValidCarbsPer100ml(food.carbs_per_100ml) || density === null) return INCOMPLETE;
    return { carbs_g: ((grams / density) * food.carbs_per_100ml) / 100, complete: true };
  }
  return INCOMPLETE;
}

function foodItemCarbs(catalog: Catalog, foodId: Id, amount: number, unit: string): CarbResult {
  const food = catalog.food(foodId);
  if (!food || !isValidAmount(amount)) return INCOMPLETE;
  const portions = catalog.portions(foodId);

  if (isVolumeUnit(unit)) {
    if (food.carbs_per_100ml != null) {
      return isValidCarbsPer100ml(food.carbs_per_100ml)
        ? { carbs_g: (amount * VOLUME_UNITS[unit] * food.carbs_per_100ml) / 100, complete: true }
        : INCOMPLETE;
    }
    // No per-100 ml basis: per-100 g + density only.
    const grams = foodAmountToGrams(amount, unit, food, portions);
    if (grams === null || food.carbs_per_100g == null || !isValidCarbsPer100g(food.carbs_per_100g)) return INCOMPLETE;
    return { carbs_g: (grams * food.carbs_per_100g) / 100, complete: true };
  }

  if (unit.startsWith(PORTION_PREFIX)) {
    const portionId = unit.slice(PORTION_PREFIX.length);
    const portion = portions.find((p) => p.id === portionId);
    if (portion == null) return INCOMPLETE;
    // carbs_g is ignored on volume portions (they only carry a weight).
    if (portion.kind !== 'volume' && portion.carbs_g != null) {
      const quantityOk = Number.isFinite(portion.quantity) && portion.quantity > 0;
      return isValidPortionCarbs(portion.carbs_g) && quantityOk
        ? { carbs_g: (amount / portion.quantity) * portion.carbs_g, complete: true }
        : INCOMPLETE;
    }
  }

  // Mass units and portions without carbs_g go through a known weight.
  const grams = foodAmountToGrams(amount, unit, food, portions);
  if (grams === null) return INCOMPLETE;
  return carbsForGrams(food, portions, grams);
}

function mealTotalCarbs(catalog: Catalog, mealId: Id, visiting: Set<Id>): CarbResult {
  if (visiting.has(mealId) || !catalog.meal(mealId)) return INCOMPLETE;
  visiting.add(mealId);
  const mealItems = catalog.mealItems(mealId);
  const total =
    mealItems.length === 0
      ? INCOMPLETE
      : sumCarbs(mealItems.map((item) => resolveItem(catalog, item.ref_type, item.ref_id, item.amount, item.unit, visiting)));
  visiting.delete(mealId);
  return total;
}

function mealItemCarbs(catalog: Catalog, mealId: Id, amount: number, unit: string, visiting: Set<Id>): CarbResult {
  const meal = catalog.meal(mealId);
  if (!meal || !Number.isFinite(amount) || amount < 0) return INCOMPLETE;
  let factor: number | null = null;
  if (unit === SERVING_UNIT && Number.isFinite(meal.yield_servings) && meal.yield_servings > 0) {
    factor = amount / meal.yield_servings;
  } else if (
    isMassUnit(unit) &&
    meal.total_weight_g != null &&
    Number.isFinite(meal.total_weight_g) &&
    meal.total_weight_g > 0
  ) {
    factor = (amount * MASS_UNITS[unit]) / meal.total_weight_g;
  }
  if (factor === null) return INCOMPLETE;
  const total = mealTotalCarbs(catalog, mealId, visiting);
  return { carbs_g: total.carbs_g * factor, complete: total.complete };
}

function resolveItem(
  catalog: Catalog,
  refType: RefType,
  refId: Id,
  amount: number,
  unit: string,
  visiting: Set<Id>,
): CarbResult {
  return refType === 'food'
    ? foodItemCarbs(catalog, refId, amount, unit)
    : mealItemCarbs(catalog, refId, amount, unit, visiting);
}

/** Carbs for one line item (a food or a meal) at the given amount and unit. */
export function itemCarbs(catalog: Catalog, refType: RefType, refId: Id, amount: number, unit: string): CarbResult {
  return resolveItem(catalog, refType, refId, amount, unit, new Set());
}

export function sumCarbs(results: CarbResult[]): CarbResult {
  return results.reduce<CarbResult>(
    (acc, r) => ({ carbs_g: acc.carbs_g + r.carbs_g, complete: acc.complete && r.complete }),
    { carbs_g: 0, complete: true },
  );
}

/** True if adding `candidateMealId` as a component of `mealId` would make a meal contain itself. */
export function wouldCreateCycle(catalog: Catalog, mealId: Id, candidateMealId: Id): boolean {
  if (candidateMealId === mealId) return true;
  const stack: Id[] = [candidateMealId];
  const seen = new Set<Id>();
  while (stack.length > 0) {
    const current = stack.pop()!;
    if (seen.has(current)) continue;
    seen.add(current);
    for (const item of catalog.mealItems(current)) {
      if (item.ref_type !== 'meal') continue;
      if (item.ref_id === mealId) return true;
      stack.push(item.ref_id);
    }
  }
  return false;
}
